### PR TITLE
Show status check for merged PRs

### PR DESCRIPTION
**Pull request by CirnoT**
This PR allows Gitea to show status check for merged PRs:
- Show status icon on PRs list for PRs whose head was removed by loading data from base repo's pull ref
- Show list of checks on PR page for merged pulls

---

chrome_2020-12-13_20-12-41

chrome_2020-12-13_20-13-00

### CODE REVIEW

**Comment by CirnoT** [comment]
> Show list of checks on PR page for merged pulls

This is pretty straightforward and I don't see any potential issues with it; basically copies what non-merged PRs do

> Show status icon on PRs list for PRs whose head was removed by loading data from base repo's pull ref

This is little more complicated because it changes how status is fetched for PR, maybe @zeripath could chime on on potential performance issues with this

File: services/pull/pull.go [hunk 1]
```diff
@@ -641,31 +642,34 @@ func GetCommitMessages(pr *models.PullRequest) string {
 
 // GetLastCommitStatus returns the last commit status for this pull request.
 func GetLastCommitStatus(pr *models.PullRequest) (status *models.CommitStatus, err error) {
-	if err = pr.LoadHeadRepo(); err != nil {
+	// load base repo
```
Comment (a1012112796) [inline services/pull/pull.go]: I think this comment is not meaningfull,suggest not add it.
```suggestion
```

**Comment by CirnoT** [comment]
Not sure this qualifies as `kind/ui` since it doesn't really touch on frontend

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
> Not sure this qualifies as `kind/ui` since it doesn't really touch on frontend

I think essentially this about showing stuff on the UI -> Therefore kind/ui!

**Comment by lunny** [comment]
The status information may be deleted when sometime?

**Review by lafriks — APPROVED** [review]

**Comment by CirnoT** [comment]
> The status information may be deleted when sometime?

We don't remove statuses from database, so unless someone were to remove it manually, no.

**Comment by lunny** [comment]
The ancient commit statuses are unuseful, especially for a big repository which has over ten thousands of commits. But they will reduce the performance of commit statuses table. So I think we should allow to delete merged or closed PR's commits statuses which haven't been merged to the git tree.

**Comment by CirnoT** [comment]
Yes the PR will work even if you were to allow removal of old statuses, it will just act as if there was no status and won't show anything (just like right now).

File: routers/repo/pull.go [hunk 2]
```diff
@@ -325,6 +325,20 @@ func PrepareMergedViewPullInfo(ctx *context.Context, issue *models.Issue) *git.C
 	}
 	ctx.Data["NumCommits"] = compareInfo.Commits.Len()
 	ctx.Data["NumFiles"] = compareInfo.NumFiles
+
+	if compareInfo.Commits.Len() > 0 {
```
Comment (6543) [inline routers/repo/pull.go]:
```suggestion
	if compareInfo.Commits.Len() != 0 {
```

File: routers/repo/pull.go [hunk 3]
```diff
@@ -325,6 +325,20 @@ func PrepareMergedViewPullInfo(ctx *context.Context, issue *models.Issue) *git.C
...
+		sha := compareInfo.Commits.Front().Value.(*git.Commit).ID.String()
+		commitStatuses, err := models.GetLatestCommitStatus(ctx.Repo.Repository, sha, 0)
+		if err != nil {
+			ctx.ServerError("GetLatestCommitStatus", err)
+			return nil
+		}
+		if len(commitStatuses) > 0 {
```
Comment (6543) [inline routers/repo/pull.go]:
```suggestion
		if len(commitStatuses) != 0 {
```

**Comment by lunny** [comment]
File conflicts

File: routers/repo/pull.go [hunk 4]
```diff
@@ -325,6 +325,20 @@ func PrepareMergedViewPullInfo(ctx *context.Context, issue *models.Issue) *git.C
 	}
 	ctx.Data["NumCommits"] = compareInfo.Commits.Len()
 	ctx.Data["NumFiles"] = compareInfo.NumFiles
+
+	if compareInfo.Commits.Len() != 0 {
+		sha := compareInfo.Commits.Front().Value.(*git.Commit).ID.String()
+		commitStatuses, err := models.GetLatestCommitStatus(ctx.Repo.Repository, sha, 0)
```
Comment (6543) [inline routers/repo/pull.go]:
```suggestion
		commitStatuses, err := models.GetLatestCommitStatus(ctx.Repo.Repository.ID, sha, models.ListOptions{})
```

**Comment by 6543** [comment]
🚀

**Pull request merged**